File: src/main/java/com/exampledigisphere/vitapetcare/admin/workDay/domain/WorkDay.java
package com.exampledigisphere.vitapetcare.admin.workDay.domain;

import com.exampledigisphere.vitapetcare.admin.timePeriod.domain.TimePeriod;
import com.exampledigisphere.vitapetcare.admin.user.domain.User;
import com.exampledigisphere.vitapetcare.config.root.BaseEntity;
import com.exampledigisphere.vitapetcare.config.root.Info;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;

import java.time.DayOfWeek;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "wrk_works_days")
@Info(
  dev = Info.Dev.heltonOliveira,
  label = Info.Label.doc,
  date = "29/12/2025",
  description = "Entidade que representa o dia de trabalho de um usuário"
)
public class WorkDay extends BaseEntity<WorkDay> {

  @Transient
  public static final String TABLE_NAME = "wrk_works_days";

  @NotNull
  @Enumerated(EnumType.STRING)
  private DayOfWeek dayOfWeek;

  @NotNull
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "user_id")
  private User user;

  @OneToMany(fetch = FetchType.LAZY, mappedBy = "workDay", cascade = CascadeType.ALL, orphanRemoval = true)
  @JsonIgnoreProperties("workDay")
  private Set<TimePeriod> shifts = new HashSet<>();

  public WorkDay() {
  }

  public DayOfWeek getDayOfWeek() {
    return dayOfWeek;
  }

  public void setDayOfWeek(DayOfWeek dayOfWeek) {
    this.dayOfWeek = dayOfWeek;
  }

  public User getUser() {
    return user;
  }

  public void setUser(User user) {
    this.user = user;
  }

  public Set<TimePeriod> getShifts() {
    return shifts;
  }

  public void setShifts(Set<TimePeriod> shifts) {
    this.shifts = shifts;
  }

  public interface Authority {
    String WORK_DAY_CREATE = "WORK_DAY_CREATE";
    String WORK_DAY_EDIT = "WORK_DAY_EDIT";
    String WORK_DAY_VIEW = "WORK_DAY_VIEW";
    String WORK_DAY_VIEW_LIST = "WORK_DAY_VIEW_LIST";
    String WORK_DAY_DELETE = "WORK_DAY_DELETE";
  }
}

File: src/main/kotlin/com/exampledigisphere/vitapetcare/admin/workDay/domain/WorkDay.kt
package com.exampledigisphere.vitapetcare.admin.workDay.domain

import com.exampledigisphere.vitapetcare.admin.timePeriod.domain.TimePeriod
import com.exampledigisphere.vitapetcare.admin.user.domain.User
import com.exampledigisphere.vitapetcare.config.root.BaseEntity
import com.exampledigisphere.vitapetcare.config.security.Permissions
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonView
import jakarta.persistence.*
import org.hibernate.Hibernate
import org.jetbrains.annotations.NotNull
import java.time.LocalDate
import java.util.Collections.emptySet

@Entity
@Table(name = "wrk_works_days")
class WorkDay : BaseEntity() {

  @field:Transient
  final val TABLE_NAME = "wrk_works_days";

  @field:NotNull
  @field:JsonView(Json.List::class)
  lateinit var date: LocalDate;

  @field:NotNull
  @field:JsonView(Json.Detail::class)
  @field:ManyToOne(fetch = FetchType.LAZY)
  lateinit var user: User

  @field:NotNull
  @field:OneToMany(fetch = FetchType.LAZY)
  @field:JsonIgnoreProperties("dayOfWeeks")
  var shifts: MutableSet<TimePeriod>? = emptySet();

  // colocar depois que criar os agendamentos
  //lateinit var  appointments: MutableList<Appointment> = mutableListOf()

  fun loadUser() = Hibernate.initialize(user);

  fun loadShifts() = Hibernate.initialize(shifts);

  interface Json {
    interface List;
    interface Detail : List;
    interface All : Detail;
  }

  object WorkDayPermissions : Permissions {
    override var resource = "WORK_DAY";
  }
}

